### DISCUSSION TITLE
Trying to update ruby from 2.6.10 to 3.1.4 (error 20240221), homebrew?

### DISCUSSION THREAD

**Discussion by sammymustafa** [rbenv/ruby-build#2355] [Build failures]
My website with a jekeyll theme suddenly stopped working and i have been trying to fix it for a while. I'm not the best at this stuff but i think "google-protobuf" needs a newer version of ruby for my website to be running again. i checked and i have ruby 2.6.10, more specifically:
ruby 2.6.10p210 (2022-04-12 revision 67958) [universal.arm64e-darwin23]

I have been trying to update it to 3.1.4 for hours to no avail. Here is the code i ran and the output, and the full build log is attached [log file.txt](https://github.com/rbenv/rbenv/files/14548728/log.file.txt).

` cd /Users/sammymustafa/Desktop/all/business/sammymustafa.github.io `
`brew install rbenv`
`rbenv install 3.1.4`
BUILD FAILED (macOS 14.4 on arm64 using ruby-build 20240221)

I think there was an issue with Homebrew refusing to force link the readline library because macOS provides its own version. With this, i updated brew and readline with certain flags and i was still getting the same failed build. This was my other attempt:

`brew update`
`brew upgrade readline`
`export LDFLAGS="-L/usr/local/opt/readline/lib"`
`export CPPFLAGS="-I/usr/local/opt/readline/include"`
`RUBY_CONFIGURE_OPTS="--with-readline-dir=$(brew --prefix readline)" rbenv install 3.1.4`

i know that after this, i need to uninstall and install google-protobuf on the ruby platform. then run bundle install on my directory to create the new gemfile.lock file [gemfile lock.txt](https://github.com/rbenv/rbenv/files/14548743/gemfile.lock.txt). please let me know what you think I need to do to make this run!! here is the theme I used for reference: https://github.com/LeNPaul/academic

any advice is greatly appreciated!!

**Comment by mislav** [comment]
> I think there was an issue with Homebrew refusing to force link the readline library because macOS provides its own version. With this, i updated brew and readline with certain flags and i was still getting the same failed build.

Hey, I looked at the log and this error is familiar to me, but I never found out what causes it:

```
compiling psych_emitter.c
readline.c:1903:37: error: use of undeclared identifier 'username_completion_function'; did you mean 'rl_username_completion_function'?
                                    rl_username_completion_function);
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                    rl_username_completion_function
readline.c:79:42: note: expanded from macro 'rl_username_completion_function'
# define rl_username_completion_function username_completion_function
                                         ^
/usr/local/opt/readline/include/readline/readline.h:494:14: note: 'rl_username_completion_function' declared here
extern char *rl_username_completion_function (const char *, int);
             ^
linking shared-object rbconfig/sizeof.bundle
1 error generated.
make[2]: *** [readline.o] Error 1
make[1]: *** [ext/readline/all] Error 2
```

This is definitely using Homebrew's readline because I see this `configure` option being passed to Ruby by rbenv install:
```
./configure <...> --with-readline-dir=/usr/local/opt/readline <...>
```

But I have no idea why this causes an error. Perhaps you could reinstall Homebrew's readline to see if it helps? Definitely search this repository's issues and discussions for previous mentions of "rl_username_completion_function". Good luck, and let us know if you solve it!
